### PR TITLE
Make exit codes a combination of meanings (aka ORable)

### PR DESCRIPTION
**Pull request by apahim**
Currently, exit codes cover all situations with one single value.

This patch makes exit codes to be ORable, so we can better represent
the variety of things that happens during an Avocado execution.

Reference: https://trello.com/c/SU5fixgH
Signed-off-by: Amador Pahim <apahim@redhat.com>

### CODE REVIEW

**Comment by ldoktor** [comment]
@apahim, @clebergnu, @adereis

A bit late, but let's question the whole idea. It requires quite a lot of values storing and propagating to get the overall status only to get slightly better notion of what happened. I don't think it adds that much value.

Take a look at the users. They:

1. Run avocado manually - they see the results, 0/1 is enough
2. Runs avocado in CI - they need machine readable results, or failure. Without the machine readable results the information that some tests were interrupted is not really helpful. They'll see it from the logs.
3. Run avocado from code - again, they parse the machine readable results.
4. Import avocado - it's not supported, but is possible. But in this case they don't need anything at all as they catch the exceptions instead.
5. Semi-automated scripts - I'm not sure if anyone does that, but theoretically one might want to re-run the job if some tests errors-out and not when tests were interrupted.Parsing the machine readable results might be an unnecessary overhead, so they can benefit from more-granular results.

So for users 1-4 we only need True/False and for user 5 we only need to report PASS, TEST_ERROR, JOB_ERROR, JOB_INTERRUPTED, GENERIC_ERROR, without the need to be or-able. The most important status is what matters

1. GENERIC_ERROR
2. JOB_ERROR
3. JOB_INTERRUPTED
4. TEST_ERROR
5. PASS

Usually I don't care if there were failures on interrupted job, it was interrupted. If I needed to know, I would be able to parse the machine readable results.

In the end I don't think we need a complex error code handling as it only slows avocado down and adds code which could be a source of potential bugs only to simplify 0-1% workflows (just guessing)

PS: I'm not a big user, but I only use `$? > 1` or and then load the `result.xml` in the CI. I do not need more.

**Comment by ldoktor** [comment]
Btw I wanted to mention, that I fully support the previous work on the runner. It should report the full status as it's convenient and part of internal API. I don't like this because it's complex with a significant footprint. For example the JobAPI would be also cope with this.

So I think we should keep the card open in the long term with a reference to this PR and wait till some users ask for it with some strong points. Avocado is quite slow and memory hungry even without this.

**Comment by adereis** [comment]
On Tue, Apr 19, 2016 at 10:33:38PM -0700, Lukáš Doktor wrote:
> Btw I wanted to mention, that I fully support the previous work
> on the runner. It should report the full status as it's
> convenient and part of internal API. I don't like this because
> it's complex with a significant footprint. For example the
> JobAPI would be also cope with this.

What's the footprint? I'm not familiar with the code, but the
diffstat looks OK (+28 -18 for the code parts). You make it sound
like this increases memory and cpu consumption in a significant
way.  How could that be true?

> 
> So I think we should keep the card open in the long term with a
> reference to this PR and wait till some users ask for it with
> some strong points. Avocado is quite slow and memory hungry
> even without this.
> 

I see this as a "nice-to-have" feature, but it doesn't have a
high priority. If it indeed adds significant complexity, then we
don't need it.

Thanks.
   - Ademar

-- 
Ademar Reis
Red Hat

^[:wq!

**Comment by clebergnu** [comment]
I have the exact opposite feeling of @ldoktor , as I explained during the last hangout. To me, it feels much more natural and less error prone to always approach the error handling by setting (ORing) the situation you're dealing with.  By doing this, there's no "race" between multiple parts of the code setting their point of view about what happened to the job.

Think of Pre/Post Job plugins, that may alter the status of the *job*... how are they to know what happened before?  Should they overwrite the status result or just add their error to a previous status that was either success or a collection of other failures?

I, and people on the mailing list also, welcome this change.

**Comment by ldoktor** [comment]
Well it's still python process, which means it's memory hungry, that is not an excuse to keep adding features, which are not really needed (nice to have, but nobody said he'd use it, something like 20MPix camera in cellphone).

So the difference is minimal, but IMO quite big compare to the benefits it brings (3runs median):

| scenario |  current rss (kb) | or-able rss (kb)| current wall time (s) | or-able wall time (s) |
| --- | --- | --- | --- | --- |
| passtest | 54364 | 54604 | 0.57 | 0.61 |
| 100x passtest | 55912 | 56408 | 02.26 | 02.28 |
| 1s interrupted | 56444| 56604 | 02.26| 02.28|
| pass/fail/int.. | 54584 | 54652| 00.64 | 00.64|
| list | 52812 | 52844| 00.24| 00.25 |

**Comment by clebergnu** [comment]
I'm against bloat, but this PoV blew me away.  Thanks for the spreadsheet, but let's think for a moment... are we developing, say hard real time systems?  If adding one extra attribute to the job class looks like problem, then the real problem is having chosen Python... Or it's time to halt everything, profile our app, and work on the bottlenecks. I can not imagine doing this kind of evaluation as a check for adding or not adding features.

**Comment by ldoktor** [comment]
As I said, if you think it's needed, fine. I just don't see many project doing it and taking benefit of it.

One possible bug could be the error status of a job interrupted by job-error, which executed some failed tests. Theoretically we should report `TEST_ERROR|JOB_ERROR` but as it crashed we have no response from the runner yet, so we don't know about `TEST_ERROR` and report only `JOB_ERROR`. It's fine by me, but users might complain.

**Comment by ldoktor** [comment]
^^ yep, I'm just comparing the pros and cons. I don't see such a big income, if others see it, we can merge it and keep improving it.

**Comment by clebergnu** [comment]
Bugs are indeed possible everywhere, and more often while handling crash scenarios. I think we can and should handle this specific case you mentioned. It's a nice heads up, and may even deserve a card to add a test for that scenario.

If anticipate a permanent breakage or major drawback with this change, let's discuss it. But as I said before, few bytes in memory should not influence our decision.

**Comment by ldoktor** [comment]
I have probably nothing more to add. I'm not this feature user. I'm not happy that it (insignificantly in the first, simple version) slows down the execution and adds lots of code which could result in bugs, resulting in even more handling and bigger overhead.

On the other hand some scenarios will be easier to define as we simply OR the status.

@clebergnu you said you know some projects with OR-able statuses. Can you please share them here? I briefly looked around and found none. (to get a notion how common this is and if people might expect this widely).

**Comment by adereis** [comment]
On Wed, Apr 20, 2016 at 06:43:34AM -0700, Lukáš Doktor wrote:
> Well it's still python process, which means it's memory hungry,
> that is not an excuse to keep adding features, which are not
> really needed (nice to have, but nobody said he'd use it,
> something like 20MPix camera in cellphone).
> 
> So the difference is minimal, but IMO quite big compare to the
> benefits it brings (3runs median):
> 
> | scenario |  current rss (kb) | or-able rss (kb)| current wall time (s) | or-able wall time (s) |
> | --- | --- | --- | --- | --- |
> | passtest | 54364 | 54604 | 0.57 | 0.61 |
> | 100x passtest | 55912 | 56408 | 02.26 | 02.28 |
> | 1s interrupted | 56444| 56604 | 02.26| 02.28|
> | pass/fail/int.. | 54584 | 54652| 00.64 | 00.64|
> | list | 52812 | 52844| 00.24| 00.25 |

Performance and memory consumption are not critical in Avocado.

So there's really no point in measuring the consequence of adding
one attribute to a class. Unless we're talking about changes that
are orders of magnitude slower, it should be basically
irrelevant.  Like a wise man once said, "premature optimization
is the root of all evils"[1].

We can and actually should profile and optimize Avocado memory
and CPU consumption from time to time, that's a much better way
to improve it.

For example, a quick cProfile run[2] shows that when running 200
passtests, of the 4.87 seconds the job takes to run, 2.11 are
spent in "sleep()" calls. If we want to optimize Avocado, that's
probably a good place to start (I would consider it very low
priority at the moment, as I don't think this is a problem).

Profiling python memory usage is a bit more complex, but there
are *tons* of opportunities if that ever becomes a concern (I've
done it in the past, when my team was developing PyQt
applications and widgets for phones back in 2010).

 1. http://c2.com/cgi/wiki?PrematureOptimization
 2. python -m cProfile -s time ./scripts/avocado run \
    $(for i in $(seq 1 200); do echo passtest; done) --silent

-- 
Ademar Reis
Red Hat

^[:wq!

**Comment by clebergnu** [comment]
Sure, I'll try to give some examples... but even if there are none out there, an application should be free to document its own exit status.  The expected contract of zero/non-zero is of course to be respected.  Other than that, IMHO, it's all fair ground.

But one *very important thing* that goes beyond the technical aspects of this discussion, is when this kind of discussion is made.  If this was added to our sprint backlog, we should all be aligned, as a team, about its relevance.  Considering it not important *after* the code was developed, that is, investment was made, is a very bad sign. We should really keep an eye on this kind of pattern.

**Comment by ldoktor** [comment]
Well that's why I said the numbers does not matter much. Handling orable statuses will always be more complex, therefor it has to consume some memory and cpu power. One can't avoid that, so I don't consider this as task for optimization, but rather as another cons point.

So from my side:

```
+ users get always all statuses (if there is no bug)
+ we don't need to think what statuses have precedence, what numbers should be bond together, etc. (all statuses are equal)
- there always will be some (minor) overhead (memory, cpu)
- not many users would really benefit from it (only for some users executing avocado using simple scripts. Advanced scripts/CI would use machine readable results and human users would use human results)
- complex handling prone to errors and if we ever spread avocado in multiple processes even greater complexity.
- quite hard selftests for these features (as we need to fake various failures and verify it set all the errors, but not other ones)
```

Overall I don't like that, but I don't think many users would met corner-cases and file bugs, so I can live with it. (the same way I'll probably have cellphone with XX MPix camera, because marketing requires XX MPix)

I see @clebergnu strongly supports it, we have some people mentioning it as "nice to have", so how about others? @apahim so far you just implemented it and had not said a word ... Do you find this feature so useful that we should add and support it? I mean we already produce 3 types of results, so yet another one is not a big deal...

**Comment by clebergnu** [comment]
ACK from me.

**Comment by adereis** [comment]
On Wed, Apr 20, 2016 at 08:20:51AM -0700, Lukáš Doktor wrote:
> Well that's why I said the numbers does not matter much. Handling orable statuses will always be more complex, therefor it has to consume some memory and cpu power. One can't avoid that, so I don't consider this as task for optimization, but rather as another cons point.
> 
> So from my side:
> 
> ```
> + users get always all statuses (if there is no bug)
> + we don't need to think what statuses have precedence, what numbers should be bond together, etc. (all statuses are equal)
> - there always will be some (minor) overhead (memory, cpu)
> - not many users would really benefit from it (only for some users executing avocado using simple scripts. Advanced scripts/CI would use machine readable results and human users would use human results)
> - complex handling prone to errors and if we ever spread avocado in multiple processes even greater complexity.
> - quite hard selftests for these features (as we need to fake various failures and verify it set all the errors, but not other ones)
> ```
> 
> Overall I don't like that, but I don't think many users would met corner-cases and file bugs, so I can live with it. (the same way I'll probably have cellphone with XX MPix camera, because marketing requires XX MPix)
> 
> I see @clebergnu strongly supports it, we have some people
> mentioning it as "nice to have", so how about others? @apahim
> so far you just implemented it and had not said a word ... Do
> you find this feature so useful that we should add and support
> it? I mean we already produce 3 types of results, so yet
> another one is not a big deal...

Per our process, we don't need (and shouldn't) vote every single
feature, as this will produce a lot of unnecessary bike shedding.

There are two possibilities once the code is submitted:

 - NACK (with a strong reason)
 - At least *one* ACK (the reviewer)

As Cleber mentioned, this kind of discussion should happen before
the code is submitted, during your weekly meetings or during
planning.

In summary, since you're not NACKn'g it and Cleber reviewed and
ACK'd it, this PR pass the criteria and should be merged.

-- 
Ademar Reis
Red Hat

^[:wq!

**Comment by ldoktor** [comment]
Agree.

**Comment by apahim** [comment]
@ldoktor I do see this feature relevance. Maybe I'm still too attached to the support processes, but I see the request, I see the 'nice-to-have' feedback, I see the value as an Avocado user, so I do support the feature implementation. As the one proposing the code, my ACK on it is implicit.
